### PR TITLE
chore: add docker image for hermetic build scripts

### PR DESCRIPTION
**Pull request by diegomarquezp**
Similar changes to those of https://github.com/googleapis/sdk-platform-java/pull/2298 by @mpeddada1 

This setup uses two triggers:
 - The first one is [library-generation-presubmit-sdk-platform-java](https://pantheon.corp.google.com/cloud-build/triggers;region=global/edit/5783744f-0820-419f-bc5e-abbbece4be0b?e=13803378&mods=monitoring_api_prod&project=cloud-devrel-kokoro-resources), which runs on each pull request. It builds a docker image with the contents of `library_generation` at HEAD and runs `library_generation/integration_tests.py` with such image
 - The second one is [library-generation-sdk-platform-java](https://pantheon.corp.google.com/cloud-build/triggers;region=global/edit/e3373892-82a2-4eac-a1f2-95523966df70?e=13803378&mods=monitoring_api_prod&project=cloud-devrel-kokoro-resources), which: 
   - is triggered upon a commit pushed to the `main` branch, then
   - builds a docker image with two tags
     - `latest`, which will be constantly updated to match the latest build
     - a tag based on the branch `${COMMIT_SHA}`
   - then the image is pushed with both tags

### tasks

- [x] create dockerfile
- [x] create cloudbuild.yaml for testing
- [x] create cloudbuild test infra
- [x] create cloudbuild.yaml for releasing the image
- [x] create cloudbuild release infra

### CODE REVIEW

**Comment by diegomarquezp** [comment]
image

File: library_generation/postprocess_library.sh [hunk 1]
```diff
@@ -22,7 +22,7 @@
 # different logic
 # 8 - configuration_yaml_path: path to the configuration yaml containing library
 # generation information for this library
-set -eo pipefail
+set -eox pipefail
```
Comment (JoeWang1127) [inline library_generation/postprocess_library.sh:25]: Can we remove `-x` here to reduce the log output?

File: library_generation/test/integration_tests.py [hunk 2]
```diff
@@ -106,11 +107,24 @@ def test_generate_repo(self):
...
 
     @classmethod
-    def __pull_repo_to(cls, dest: Path, repo: str, committish: str):
-        repo_url = f"{repo_prefix}/{repo}"
-        print(f"Cloning repository {repo_url}")
-        repo = Repo.clone_from(repo_url, dest)
+    def __pull_repo_to(cls, default_dest: Path, repo: str, committish: str) -> str:
+        if "RUNNING_IN_DOCKER" in os.environ:
+            # the docker image expects the repo to be in /workspace
+            dest_str = "/workspace"
```
Comment (JoeWang1127) [inline library_generation/test/integration_tests.py:113]:
```suggestion
            dest_in_docker = "/workspace"
```

File: library_generation/test/integration_tests.py [hunk 3]
```diff
@@ -106,11 +107,24 @@ def test_generate_repo(self):
...
+            run_process_and_print_output(
+                ["git", "config", "--global", "--add", "safe.directory", dest_str],
+                "Add /workspace to safe directories",
+            )
+            dest = Path(dest_str)
+            repo = Repo(dest)
```
Comment (JoeWang1127) [inline library_generation/test/integration_tests.py:119]:
```suggestion
            repo = Repo(Path(dest_in_docker))
```

File: library_generation/test/integration_tests.py (hunk 2 above)
Comment (JoeWang1127) [reply library_generation/test/integration_tests.py:113]: nit.

File: library_generation/postprocess_library.sh [hunk 4]
```diff
@@ -61,22 +61,47 @@ fi
...
+repo_binding="${postprocessing_target}"
+repo_workspace="/repo"
+preprocessed_libraries_binding="${owlbot_cli_source_folder}"
+
+# When running docker inside docker, we run into the issue of volume bindings
+# being mapped from the host machine to the child container (instead of the
+# parent container to child container) because we bind the `docker.sock` socket
+# to the parent container (i.e. docker calls use the host's filesystem context)
+# We solve this by referencing environment variables that will be
+# set to produce the correct volume mapping.
```
Comment (JoeWang1127) [inline library_generation/postprocess_library.sh:80]: Good finding.

Could you add the source of this finding?

**Comment by JoeWang1127** [comment]
Thanks for this PR, @diegomarquezp. LGTM.

As a follow up PR, could you modify the [nightly CI](https://github.com/googleapis/google-cloud-java/blob/main/.github/workflows/verify-generation-config.yaml) in google-cloud-java to use the image, rather than copying script from this repository?

**Review by JoeWang1127 — APPROVED** [review]

**Comment by diegomarquezp** [comment]
> Thanks for this PR, @diegomarquezp. LGTM.
> 
> As a follow up PR, could you modify the [nightly CI](https://github.com/googleapis/google-cloud-java/blob/main/.github/workflows/verify-generation-config.yaml) in google-cloud-java to use the image, rather than copying script from this repository?

@JoeWang1127 We will have to wait for it to be published when the next release PR is merged, if that's ok. Otherwise I can modify the cloud build workflow to manually create one. WDYT?

File: .cloudbuild/library_generation.Dockerfile [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+# install tools
+RUN apt-get update && apt-get install -y \
+	unzip openjdk-17-jdk rsync maven \
+	&& apt-get clean
+
+COPY library_generation /src
+
+RUN rm $(which python3)
+RUN ln -s $(which python3.11) /usr/local/bin/python
+RUN ln -s $(which python3.11) /usr/local/bin/python3
```
Comment (JoeWang1127) [inline .cloudbuild/library_generation.Dockerfile:27]: Why do we need these steps to set up python? I thought the python has already been set up in the base image.

Comment (diegomarquezp) [reply .cloudbuild/library_generation.Dockerfile:27]: yes, there are several version of python (e.g. `/usr/local/bin/python3.9`). The default `python3` and `python` point to `python3.12`. This step basically creates a symbolic link to `python3.11`

**Comment by diegomarquezp** [comment]
> > @JoeWang1127 We will have to wait for it to be published when the next release PR is merged, if that's ok. Otherwise I can modify the cloud build workflow to manually create one. WDYT?
> 
> Can we trigger a cloud build whenever there's a new commit to `library_generation` directory? We can use the image to test how the scripts work in google-cloud-java. We only have one release per two weeks which is slow for us to verify the scripts.
> 
> Also, could you tag the image with `latest` when the commit is merged, so that we don't need to change which image we are using in google-cloud-java.

@JoeWang1127 I added a special trigger to build and publish a docker image based on pushes to the `main` branch of sdk-platform-java

image

Both the `latest` tag and a tag based on the commit SHA of sdk-platform-java are applied

File: library_generation/test/container_integration_tests.sh [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+fi
+pushd google-cloud-java
+git reset --hard main
+popd
+if [[ $(docker volume inspect repo) != '[]' ]]; then
+  docker volume rm repo
+fi
+docker volume create --name "repo" --opt "type=none" --opt "device=$(pwd)/google-cloud-java" --opt "o=bind"
+
+image_id="gcr.io/cloud-devrel-public-resources/java-library-generation:${SHARED_DEPENDENCIES_VERSION}"
```
[the remark this answers is not in the document]
Comment (diegomarquezp) [reply library_generation/test/container_integration_tests.sh]: Sounds good. I added `IMAGE_ID` as a passed env var

File: library_generation/test/container_integration_tests.sh [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+if [[ ! -d google-cloud-java ]]; then
+  git clone https://github.com/googleapis/google-cloud-java
+fi
+pushd google-cloud-java
+git reset --hard main
+popd
+if [[ $(docker volume inspect repo) != '[]' ]]; then
+  docker volume rm repo
+fi
+docker volume create --name "repo" --opt "type=none" --opt "device=$(pwd)/google-cloud-java" --opt "o=bind"
```
[the remark this answers is not in the document]
Comment (diegomarquezp) [reply library_generation/test/container_integration_tests.sh:29]: Added a comment explaining why we create this volume

**Comment by diegomarquezp** [comment]
@suztomo do you agree with pushing the image to the [`cloud-devrel-public-resources` registry](https://pantheon.corp.google.com/gcr/images/cloud-devrel-public-resources/global/java-library-generation?e=13803378&mods=monitoring_api_prod&project=cloud-devrel-public-resources)? We can also use `cloud-devrel-kokoro-resources`

cc: @mpeddada1

**Comment by diegomarquezp** [comment]
> Thanks for this PR, @diegomarquezp. LGTM.
> 
> As a follow up PR, could you modify the [nightly CI](https://github.com/googleapis/google-cloud-java/blob/main/.github/workflows/verify-generation-config.yaml) in google-cloud-java to use the image, rather than copying script from this repository?

I switched to use the docker image in https://github.com/googleapis/google-cloud-java/pull/10435

**Comment by suztomo** [comment]
Yes, cloud-devrel-public-resources is simpler. No authentication when using it.

**Pull request merged**